### ISSUE TITLE
Feature request: support for RSS feeds

### ISSUE THREAD

**Issue by infokiller** [burtonator/polar-bookshelf#994]
**Is your feature request related to a problem? Please describe.**
I'm using a content aggregator (feed reader), specifically Inoreader, and in order to read the content from it in Polar I need to manually transfer web page URLs.

**Describe the solution you'd like**
Integration with RSS feeds for automatically adding documents to Polar.

**Describe alternatives you've considered**
- Manually copying URLs of articles from my content aggregator to Polar
- Writing a script for importing the URLs of articles from Inoreader to the Polar database

**Comment by burtonator** [comment]
Polar would need to support a way to automatically capture a URL and this is very very difficult to do reliably due to HTML issues.  This might take 2-6 months to see implemented FYI.

We might have support for some basic features sooner like simpler capture which need to be done first.

**Comment by infokiller** [comment]
Thanks for the quick response!
Can you please elaborate on the reliability issue related to HTML?

**Comment by burtonator** [comment]
I wrote up a more complete blog post here:

https://getpolarized.io/2019/05/11/portable-web-documents.html

The main issues are that a lot of pages can't just be transparently captured.  But you can read about it there..  

In short, there's not really the concept of a page having been 'loaded' due to Javascript.

**Comment by burtonator** [comment]
Also, I think this would only be possible in the desktop version - not the web version.  It would have to create 'offscreen' windows to render the content and store ad phz/epub

**Comment by infokiller** [comment]
If I understand the issue correctly, it seems that capturing the document from URLs is indeed very hard if you want all the features of the original document, but I would think that "only" extracting the major content (basically text and images) can be done reasonably well for most of the content online. A few example implementations of the functionality I'm referring to:
- [Chrome's DOM distiller](https://github.com/chromium/dom-distiller)
- https://outline.com
- https://getpocket.com
- Many RSS clients that have show articles inline such as Inoreader

I'd wager that "capturing" the URLs in this manner will suffice for the vast majority of users who are interested in this functionality.

**Comment by infokiller** [comment]
Do you have an estimate on the effort it would take to integrate this functionality into Polar for a first time contributor?
And would you be open to merging such a PR?

Thanks

**Comment by burtonator** [comment]
@infokiller I think it would require the following.

I'm fine merging it if the code is clean. Would need to use typescript and the fetch API to fetch parse the content.  The 'Datastore' API can be used to add the documents.

Some type of concept of 'channels' for feeds so that in the sidebar the user can pick a specific feed.

We can probably use tags for this.  Tags support hierarchy and I imagine we could finish up support for this soon. I'm about to add support for folders too so maybe there's some overlap here.

We don't have an API just yet for creating a captured PHZ document from raw HTML but I think we could put one together somewhat easily.

The biggest challenge though is that we might have to update our 'background task API' which isn't as flushed out as I would like.

This is designed so that Polar can do background tasks and show the user what's going on.  

The idea is that async IO tasks would all go into this system to inform the user of the status and their progress.

We will also need the ability to cancel the tasks too, schedule them, etc.  

We should also probably use the new chrome network bandwidth API to detect when the user is on a 4G network to avoid fetchng RSS feeds while mobile or offline.

I think the UI and change for this would be the hardest to merge honestly and maybe we can just have that as a separate task since that's somewhat isolated.

Would require tests too and maybe we need to find a robust RSS parsing library.

**Comment by burtonator** [comment]
And hopefully once we get more of the core internals of Polar tightened up that the RSS and feed support will be really simple to add.

**Comment by infokiller** [comment]
Thanks for the thorough explanation!

I'm fine with doing it in Typescript, using the fetch API, and adding tests (not familiar with the Datastore API but shouldn't be an issue).

The UI side indeed sounds like the hardest thing here, and this is the area where I would be least helpful too. My shitty web UI skills will probably suffice for adding a basic POC UI, but not something complicated or very polished.

BTW, I think that even only adding the option for a user to import an OPML file and add a button for "Fetch RSS feeds" will already be useful for people like me. And then later we can add periodic fetching without the need to press that button every time. So if that simplifies things significantly, I think we should also consider it.

Regarding the channels UI, I agree that tags should work here. So for `https://example.com/feed` we would add a tag such as `rss:https://example.com/feed`. If tags support hierarchy that's even better because all the feeds can be children of the `feeds` tag or whatever.

**Comment by burtonator** [comment]
Maybe if you could find us a good RSS parsing infra first that might be a good first start.

I'm actually a domain expert here as I was one of the co-inventors of RSS and ran a search company for the past decade that indexed petabytes of RSS.  

... the key part is going to be the parser as there are about 15 formats including Atom.

I might be able to do the UI and then advise on the backend technologies here.  

OK.. so you're saying just have manual fetching for now?  Yeah. Probably the best way to simplify it I would think. 

The way I was envisioning it is that the 'feeds' would be elsewhere and the documents they fetched would get added to the document repository.  So if you added ArsTechnica feed nothing would show up until you fetched it for the first time at which point you would have about 15 documents in the doc repo.

We might want them to automatically get tags that are assigned to the feed.  

We're going to support folders soon so that can be done too.

I think that for now this would ONLY work on the desktop version.  

... but btw. Doing this in node + fetch means we COULD schedule this in firebase cloud functions via cron at some point if we wanted too but I like the idea of doing it on the users workstation if we keep the load reasonable. 

I think I have an AsyncQueue impl that has rate limiting implemented. We could also implement a TokenBucketFilter and give it a policy so that it can only fetch a document every 5 minutes or so which would prevent the users machine from being overwhelmed.

**Comment by burtonator** [comment]
and yeah.. we can just start building it out on the backend with unit tests and architecture and then add it into the UI when we are ready.

**Comment by burtonator** [comment]
There's this ...

https://github.com/danmactough/node-feedparser

**Comment by infokiller** [comment]
> Maybe if you could find us a good RSS parsing infra first that might be a good first start.
> 
> I'm actually a domain expert here as I was one of the co-inventors of RSS and ran a search company for the past decade that indexed petabytes of RSS.
> 
> ... the key part is going to be the parser as there are about 15 formats including Atom.
Right. I'll try to carve out some time for this.
> 
> I might be able to do the UI and then advise on the backend technologies here.
> 
> OK.. so you're saying just have manual fetching for now? Yeah. Probably the best way to simplify it I would think.
> 
> The way I was envisioning it is that the 'feeds' would be elsewhere and the documents they fetched would get added to the document repository. So if you added ArsTechnica feed nothing would show up until you fetched it for the first time at which point you would have about 15 documents in the doc repo.

Makes sense to separate the two- they should be orthogonal features.
To further simplify things on the UI side, we could start by representing the feeds in the backend using an OPML file. Users will be able to edit this file to configure their feeds. This will allow advanced users to export the OPML file from their feed aggregator and then add it to Polar. I think many RSS users are more "advanced" than the average web user, so this might be ok to check if there's a demand for the RSS feature.
Obviously, if this will be released in Polar before everything is polished, the release announcement will have to clarify this is very experimental. But it may still be useful to start getting feedback and understand how many users want this.
> 
> We might want them to automatically get tags that are assigned to the feed.
Do you mean having rules for tagging documents coming from feeds?
> 
> We're going to support folders soon so that can be done too.
> 
> I think that for now this would ONLY work on the desktop version.

Yes, I think this is fine at this point.
> 
> ... but btw. Doing this in node + fetch means we COULD schedule this in firebase cloud functions via cron at some point if we wanted too but I like the idea of doing it on the users workstation if we keep the load reasonable.
> 
> I think I have an AsyncQueue impl that has rate limiting implemented. We could also implement a TokenBucketFilter and give it a policy so that it can only fetch a document every 5 minutes or so which would prevent the users machine from being overwhelmed.

**Comment by infokiller** [comment]
> There's this ...
> 
> [danmactough/node-feedparser](https://github.com/danmactough/node-feedparser)

Other options I found from a quick search:
- https://github.com/fent/feedme.js
- https://github.com/bobby-brennan/rss-parser (only supports RSS)

There's also https://github.com/alabeduarte/feedparser-promised for using promises with feedparser.
And [here's some stats from npm](https://npmcompare.com/compare/feedme,feedparser,rss-parser) comparing these libraries.

**Comment by AdamCanady** [comment]
Any update on this or other options for automatically dumping URLs into polar?

**Comment by AdamCanady** [comment]
I found the capture.ts script and spent a couple hours trying to get it up and running but have so far been unsuccessful.

I detailed some of what I tried here: https://github.com/burtonator/polar-app-public/issues/2
